src/status.ts

```diff
@@ -0,0 +1,278 @@
+  private selector: vscode.DocumentSelector;
+  private serverStatus: vscode.LanguageStatusItem;
+  private yjitStatus: vscode.LanguageStatusItem;
+
+  constructor(context: vscode.ExtensionContext, ruby: Ruby) {
+    this.context = context;
+    this.ruby = ruby;
+    this.selector = {
+      scheme: "file",
+      language: "ruby",
+    };
+
+    this.serverStatus = this.createStatusItem(
+      "serverStatus",
+      "Ruby LSP: Starting...",
+      vscode.LanguageStatusSeverity?.Information
+    );
+
+    this.serverStatus.command = {
+      title: "Restart server",
+      command: "rubyLsp.serverOptions",
+    };
+
+    this.yjitStatus = vscode.languages.createLanguageStatusItem(
+      "yjit",
+      this.selector
+    );
+
+    this.createYjitStatus(this.ruby);
+    this.createRubyStatus(this.ruby);
+
+    this.registerCommands();
+  }
+
+  public async updateStatus(status: ServerCommand) {
+    this.serverStatus.severity = vscode.LanguageStatusSeverity?.Information;
+
+    switch (status) {
+      case ServerCommand.Start: {
+        this.serverStatus.text = "Ruby LSP: Running...";
+        this.serverStatus.command!.title = "Stop Server";
+
+        this.activateGemOutdatedButton();
+
+        break;
+      }
+      case ServerCommand.Stop: {
+        this.serverStatus.text = "Ruby LSP: Stopped";
+        this.serverStatus.command!.title = "Start Server";
+
+        break;
+      }
+      case ServerCommand.Restart: {
+        this.serverStatus.text = "Ruby LSP: Error";
+        this.serverStatus.severity = vscode.LanguageStatusSeverity?.Error;
+        this.serverStatus.command!.title = "Restart Server";
+      }
+    }
+  }
+
+  public async installGems(): Promise<boolean> {
+    if (!(await Bundler.bundleCheck())) {
+      this.serverStatus.text = "Ruby LSP: Error";
+      this.serverStatus.severity = vscode.LanguageStatusSeverity?.Error;
+
+      const status: vscode.LanguageStatusItem = this.createStatusItem(
+        "installGems",
+        "Ruby LSP: The gems in the bundle are not installed.",
+        vscode.LanguageStatusSeverity?.Error
+      );
+
+      const commandId = "rubyLsp.installGems";
+
+      this.context.subscriptions.push(
+        vscode.commands.registerCommand(commandId, () => {
+          status.text = "Ruby LSP: Installing gems...";
+          status.busy = true;
+          Bundler.bundleInstall()
+            .then(() => {
+              status.dispose();
+              this.addMissingGem();
+            })
+            .catch(() => {
+              status.dispose();
+              this.createStatusItem(
+                "installFail",
+                "Ruby LSP: Failed to install gems.",
+                vscode.LanguageStatusSeverity?.Error
+              );
+            });
+        })
+      );
+
+      status.command = {
+        title: "Run bundle install",
+        command: commandId,
+      };
+      return true;
+    }
+    return false;
+  }
+
+  public async addMissingGem(): Promise<boolean> {
```

Comment:
Not a blocker as this is a matter of preference, but both `addMissingGem` and `installGems` could use an [early return](https://gomakethings.com/the-early-return-pattern-in-javascript/) to avoid having most of the function body be inside of an `if` statement.